worker/tracing/tracer.go

```diff
@@ -0,0 +1,213 @@
+	"go.opentelemetry.io/otel/trace"
+	"gopkg.in/tomb.v2"
+
+	coretracing "github.com/juju/juju/core/tracing"
+	"github.com/juju/juju/version"
+)
+
+type tracer struct {
+	tomb tomb.Tomb
+
+	namespace          string
+	client             otlptrace.Client
+	clientProvider     *sdktrace.TracerProvider
+	clientTracer       trace.Tracer
+	stackTracesEnabled bool
+	logger             Logger
+}
+
+// NewTracerWorker returns a new tracer worker.
+func NewTracerWorker(
+	ctx context.Context,
+	namespace string,
+	endpoint string,
+	insecureSkipVerify bool,
+	stackTracesEnabled bool,
+	logger Logger,
+) (TrackedTracer, error) {
+	client, clientProvider, clientTracer, err := newClient(ctx, namespace, endpoint, insecureSkipVerify)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	t := &tracer{
+		namespace:          namespace,
+		client:             client,
+		clientProvider:     clientProvider,
+		clientTracer:       clientTracer,
+		stackTracesEnabled: stackTracesEnabled,
+		logger:             logger,
+	}
+
+	t.tomb.Go(t.loop)
+	return t, nil
+}
+
+// Kill implements the worker.Worker interface.
+func (t *tracer) Kill() {
+	t.tomb.Kill(nil)
+}
+
+// Wait implements the worker.Worker interface.
+func (t *tracer) Wait() error {
+	return t.tomb.Wait()
+}
+
+// Start creates a span and a context.Context containing the newly-created span.
+func (t *tracer) Start(ctx context.Context, name string, opts ...coretracing.Option) (context.Context, coretracing.Span) {
+	o := coretracing.NewTracerOptions()
+	for _, opt := range opts {
+		opt(o)
+	}
+
+	// Tie the lifetime of the span to the lifetime of the worker. If the
+	// worker dies then the span will be ended. The consumer of the span
+	// should use the context returned from this method to ensure that the
+	// they also die at the same time as the worker.
+	var (
+		cancel context.CancelFunc
+		span   trace.Span
+	)
+	ctx, cancel = t.scopedContext(ctx)
+
+	// Grab any attributes from the options and add them to the span.
+	attrs := attributes(o.Attributes())
+	ctx, span = t.clientTracer.Start(ctx, name, trace.WithAttributes(attrs...))
+
+	if t.logger.IsTraceEnabled() {
+		spanContext := span.SpanContext()
+		t.logger.Tracef("SpanContext: span-id %s, trace-id %s", spanContext.SpanID(), spanContext.TraceID())
+	}
+
+	return ctx, &managedSpan{
+		span:               span,
+		cancel:             cancel,
+		stackTracesEnabled: t.requiresStackTrace(o.StackTrace()),
+	}
+}
+
+// requiresStackTrace returns true if the stack trace should be enabled on the
+// span or if the stack trace is enabled on the tracer (globally).
+func (t *tracer) requiresStackTrace(spanStackTrace bool) bool {
+	if spanStackTrace {
+		return true
+	}
+	return t.stackTracesEnabled
+}
+
+func (t *tracer) loop() error {
+	defer func() {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+		defer cancel()
+
+		if err := t.client.Stop(ctx); err != nil {
+			t.logger.Infof("failed to stop client: %v", err)
```

Comment:
Are these worthy of `Errorf`?

Reply:
Error indicates you should perform an action when this happens, nothing we can do here.